waku/v2/protocol/waku_noise/noise.nim

```diff
@@ -50,45 +84,166 @@ proc randomSeqByte*(rng: var BrHmacDrbgContext, size: uint32): seq[byte] =
 # It takes a Cipher State (with key, nonce, and associated data) and encrypts a plaintext
+# The cipher state in not changed
 proc encrypt*(
     state: ChaChaPolyCipherState,
     plaintext: openArray[byte]): ChaChaPolyCiphertext
-    {.noinit, raises: [Defect].} =
-  #TODO: add padding
+    {.noinit, raises: [Defect, NoiseEmptyChaChaPolyInput].} =
+  #If plaintext is empty, we raise an error
+  if plaintext == @[]:
+    raise newException(NoiseEmptyChaChaPolyInput, "Tried to encrypt empty plaintext")
   var ciphertext: ChaChaPolyCiphertext
+  # Since ChaChaPoly's library "encrypt" primitive directly changes the input plaintext to the ciphertext,
+  # we copy the plaintext into the ciphertext variable and we pass the latter to encrypt
   ciphertext.data.add plaintext
+  #TODO: add padding
+  # ChaChaPoly.encrypt takes as input: the key (k), the nonce (nonce), a data structure for storing the computed authorization tag (tag), 
+  # the plaintext (overwritten to ciphertext) (data), the associated data (ad)
   ChaChaPoly.encrypt(state.k, state.nonce, ciphertext.tag, ciphertext.data, state.ad)
   return ciphertext
 
 # ChaChaPoly decryption
 # It takes a Cipher State (with key, nonce, and associated data) and decrypts a ciphertext
+# The cipher state is not changed
 proc decrypt*(
     state: ChaChaPolyCipherState, 
     ciphertext: ChaChaPolyCiphertext): seq[byte]
-    {.raises: [Defect, NoiseDecryptTagError].} =
+    {.raises: [Defect, NoiseEmptyChaChaPolyInput, NoiseDecryptTagError].} =
+  #If plaintext is empty, we raise an error
+  if ciphertext.data == @[]:
+    raise newException(NoiseEmptyChaChaPolyInput, "Tried to decrypt empty ciphertext")
   var
+    # The input authorization tag
     tagIn = ciphertext.tag
+    # The authorization tag computed during decryption
     tagOut: ChaChaPolyTag
+  # Since ChaChaPoly's library "decrypt" primitive directly changes the input ciphertext to the plaintext,
+  # we copy the ciphertext into the plaintext variable and we pass the latter to decrypt
   var plaintext = ciphertext.data
+  # ChaChaPoly.decrypt takes as input: the key (k), the nonce (nonce), a data structure for storing the computed authorization tag (tag), 
+  # the ciphertext (overwritten to plaintext) (data), the associated data (ad)
   ChaChaPoly.decrypt(state.k, state.nonce, tagOut, plaintext, state.ad)
   #TODO: add unpadding
   trace "decrypt", tagIn = tagIn.shortLog, tagOut = tagOut.shortLog, nonce = state.nonce
+  # We check if the authorization tag computed while decrypting is the same as the input tag
   if tagIn != tagOut:
     debug "decrypt failed", plaintext = shortLog(plaintext)
     raise newException(NoiseDecryptTagError, "decrypt tag authentication failed.")
   return plaintext
 
-# Generates a random Cipher Test for testing encryption/decryption
+# Generates a random ChaChaPoly Cipher State for testing encryption/decryption
 proc randomChaChaPolyCipherState*(rng: var BrHmacDrbgContext): ChaChaPolyCipherState =
```

Comment:
Yes, I thought of moving it somewhere else, but then I'm forced to make public/accessible the fields in a `ChaChaPolyCipherState`, which technically should be avoided since contains key material (although in current commit they are public, but they were so just for testing: I'll make them private in next commit).

Reply:
Made fields private in https://github.com/status-im/nim-waku/commit/b4f0c5447942f020bad4d7f90489574008be98c5. I kept randomChaChaPolyCiphertState in noise.nim